File: internal/api/network.go
package api

import (
	"context"
	"encoding/json"
	"net"
	"net/http"
	"sync"
	"time"

	"github.com/pxelab/pxelab/internal/network"
)

type NetworkHandler struct{}

func (h *NetworkHandler) Ping(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Host      string `json:"host"`
		Count     int    `json:"count"`
		TimeoutMs int    `json:"timeout_ms"`
		IntervalMs int   `json:"interval_ms"`
		Size      int    `json:"size"`
		TTL       int    `json:"ttl"`
		Interface string `json:"interface"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		Error(w, http.StatusBadRequest, "无效的请求体")
		return
	}
	if req.Host == "" {
		Error(w, http.StatusBadRequest, "主机地址不能为空")
		return
	}

	// Resolve interface to local IP
	var localIP string
	if req.Interface != "" {
		localIP = resolveInterfaceIP(req.Interface)
	}

	opts := network.PingOptions{
		Count:     req.Count,
		Timeout:   time.Duration(req.TimeoutMs) * time.Millisecond,
		Interval:  time.Duration(req.IntervalMs) * time.Millisecond,
		Size:      req.Size,
		TTL:       req.TTL,
		LocalAddr: localIP,
	}

	result, err := network.Ping(r.Context(), req.Host, opts, nil)
	if err != nil {
		Error(w, http.StatusInternalServerError, err.Error())
		return
	}
	OK(w, result)
}

// POST /api/v1/network/ping/stream — SSE continuous ping
func (h *NetworkHandler) PingStream(w http.ResponseWriter, r *http.Request) {
	flusher, ok := w.(http.Flusher)
	if !ok {
		Error(w, http.StatusInternalServerError, "不支持 SSE")
		return
	}

	var req struct {
		Host      string `json:"host"`
		Count     int    `json:"count"`
		TimeoutMs int    `json:"timeout_ms"`
		IntervalMs int   `json:"interval_ms"`
		Size      int    `json:"size"`
		TTL       int    `json:"ttl"`
		Interface string `json:"interface"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		Error(w, http.StatusBadRequest, "无效的请求体")
		return
	}
	if req.Host == "" {
		Error(w, http.StatusBadRequest, "主机地址不能为空")
		return
	}

	var localIP string
	if req.Interface != "" {
		localIP = resolveInterfaceIP(req.Interface)
	}

	opts := network.PingOptions{
		Count:     req.Count,
		Timeout:   time.Duration(req.TimeoutMs) * time.Millisecond,
		Interval:  time.Duration(req.IntervalMs) * time.Millisecond,
		Size:      req.Size,
		TTL:       req.TTL,
		LocalAddr: localIP,
	}

	w.Header().Set("Content-Type", "text/event-stream")
	w.Header().Set("Cache-Control", "no-cache")
	w.Header().Set("Connection", "keep-alive")

	ctx, cancel := context.WithCancel(r.Context())
	defer cancel()

	go func() {
		<-r.Context().Done()
		cancel()
	}()

	var wg sync.WaitGroup
	wg.Add(1)

	result, err := network.Ping(ctx, req.Host, opts, func(pkt network.PingPacket) {
		data, _ := json.Marshal(pkt)
		select {
		case <-ctx.Done():
			return
		default:
			_, _ = w.Write([]byte("data: " + string(data) + "\n\n"))
			flusher.Flush()
		}
	})

	wg.Done()

	// 流式执行失败时通知前端一个 error 帧，避免前端永远 loading
	if err != nil {
		writeStreamError(w, flusher, r.Context(), err)
		return
	}

	if result != nil {
		summary, _ := json.Marshal(map[string]any{
			"type":     "summary",
			"ip":       result.IP,
			"sent":     result.Sent,
			"received": result.Received,
			"lost":     result.Lost,
			"min_rtt":  result.MinRTT,
			"max_rtt":  result.MaxRTT,
			"avg_rtt":  result.AvgRTT,
		})
		_, _ = w.Write([]byte("data: " + string(summary) + "\n\n"))
		flusher.Flush()
	}
}

func (h *NetworkHandler) Traceroute(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Host      string `json:"host"`
		MaxHops   int    `json:"max_hops"`
		TimeoutMs int    `json:"timeout_ms"`
		Probes    int    `json:"probes"`
		Interface string `json:"interface"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		Error(w, http.StatusBadRequest, "无效的请求体")
		return
	}
	if req.Host == "" {
		Error(w, http.StatusBadRequest, "主机地址不能为空")
		return
	}

	var localIP string
	if req.Interface != "" {
		localIP = resolveInterfaceIP(req.Interface)
	}

	opts := network.TracerouteOptions{
		MaxHops:   req.MaxHops,
		Timeout:   time.Duration(req.TimeoutMs) * time.Millisecond,
		Probes:    req.Probes,
		LocalAddr: localIP,
	}

	result, err := network.Traceroute(r.Context(), req.Host, opts, nil)
	if err != nil {
		Error(w, http.StatusInternalServerError, err.Error())
		return
	}
	OK(w, result)
}

// POST /api/v1/network/traceroute/stream — SSE 逐跳实时输出
func (h *NetworkHandler) TracerouteStream(w http.ResponseWriter, r *http.Request) {
	flusher, ok := w.(http.Flusher)
	if !ok {
		Error(w, http.StatusInternalServerError, "不支持 SSE")
		return
	}

	var req struct {
		Host      string `json:"host"`
		MaxHops   int    `json:"max_hops"`
		TimeoutMs int    `json:"timeout_ms"`
		Probes    int    `json:"probes"`
		Interface string `json:"interface"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		Error(w, http.StatusBadRequest, "无效的请求体")
		return
	}
	if req.Host == "" {
		Error(w, http.StatusBadRequest, "主机地址不能为空")
		return
	}

	var localIP string
	if req.Interface != "" {
		localIP = resolveInterfaceIP(req.Interface)
	}

	opts := network.TracerouteOptions{
		MaxHops:   req.MaxHops,
		Timeout:   time.Duration(req.TimeoutMs) * time.Millisecond,
		Probes:    req.Probes,
		LocalAddr: localIP,
	}

	w.Header().Set("Content-Type", "text/event-stream")
	w.Header().Set("Cache-Control", "no-cache")
	w.Header().Set("Connection", "keep-alive")

	ctx, cancel := context.WithCancel(r.Context())
	defer cancel()
	go func() {
		<-r.Context().Done()
		cancel()
	}()

	result, err := network.Traceroute(ctx, req.Host, opts, func(hop network.TracerouteHop) {
		data, _ := json.Marshal(hop)
		select {
		case <-ctx.Done():
			return
		default:
			_, _ = w.Write([]byte("data: " + string(data) + "\n\n"))
			flusher.Flush()
		}
	})
	if err != nil {
		writeStreamError(w, flusher, r.Context(), err)
		return
	}
	if result != nil {
		summary, _ := json.Marshal(map[string]any{
			"type": "summary",
			"host": result.Host,
			"ip":   result.IP,
			"hops": len(result.Hops),
		})
		_, _ = w.Write([]byte("data: " + string(summary) + "\n\n"))
		flusher.Flush()
	}
}

// writeStreamError 向 SSE 流发送一个 error 帧；客户端已断开时跳过
func writeStreamError(w http.ResponseWriter, flusher http.Flusher, ctx context.Context, err error) {
	if ctx.Err() != nil {
		return
	}
	data, _ := json.Marshal(map[string]any{"type": "error", "message": err.Error()})
	_, _ = w.Write([]byte("data: " + string(data) + "\n\n"))
	flusher.Flush()
}

// GET /api/v1/network/interfaces — 返回可用网卡（名称 + IP）
func (h *NetworkHandler) ListInterfaces(w http.ResponseWriter, r *http.Request) {
	type ifaceInfo struct {
		Name string   `json:"name"`
		IPs  []string `json:"ips"`
	}
	var result []ifaceInfo
	ifaces, err := net.Interfaces()
	if err != nil {
		OK(w, result)
		return
	}
	for _, iface := range ifaces {
		if iface.Flags&net.FlagUp == 0 || iface.Flags&net.FlagLoopback != 0 {
			continue
		}
		addrs, err := iface.Addrs()
		if err != nil {
			continue
		}
		info := ifaceInfo{Name: iface.Name}
		for _, addr := range addrs {
			var ip net.IP
			switch v := addr.(type) {
			case *net.IPNet:
				ip = v.IP
			case *net.IPAddr:
				ip = v.IP
			}
			if ip == nil || ip.IsLoopback() {
				continue
			}
			if ip.To4() != nil {
				info.IPs = append(info.IPs, ip.String())
			}
		}
		if len(info.IPs) > 0 {
			result = append(result, info)
		}
	}
	OK(w, result)
}

func resolveInterfaceIP(name string) string {
	ifaces, err := net.Interfaces()
	if err != nil {
		return ""
	}
	for _, iface := range ifaces {
		if iface.Name != name {
			continue
		}
		addrs, err := iface.Addrs()
		if err != nil {
			continue
		}
		for _, addr := range addrs {
			var ip net.IP
			switch v := addr.(type) {
			case *net.IPNet:
				ip = v.IP
			case *net.IPAddr:
				ip = v.IP
			}
			if ip != nil && ip.To4() != nil {
				return ip.String()
			}
		}
	}
	return ""
}

File: web/src/api/network.ts
import type { ApiResponse } from './types'
import { getBaseURL, getSessionToken, request } from './http'

// ── Network Diagnostics ──
export interface PingOptions {
  host: string
  count?: number
  timeout_ms?: number
  interval_ms?: number
  size?: number
  ttl?: number
  interface?: string
}

export interface PingPacket {
  seq: number
  rtt: number
  ttl: number
  bytes: number
  error?: string
}

export interface PingResult {
  host: string
  ip: string
  sent: number
  received: number
  lost: number
  min_rtt: number
  max_rtt: number
  avg_rtt: number
  stddev_rtt: number
  packets: PingPacket[]
  reachable: boolean
  error?: string
}

export interface TracerouteOptions {
  host: string
  max_hops?: number
  timeout_ms?: number
  probes?: number
  interface?: string
}

export interface TracerouteHop {
  ttl: number
  ip: string
  host: string
  rtt: number
  rtts: number[]
  timeout: boolean
  avg_rtt?: number
}

export interface TracerouteResult {
  host: string
  ip: string
  hops: TracerouteHop[]
  error?: string
}

export interface NetworkInterface {
  name: string
  ips: string[]
}

export function networkPing(options: PingOptions): Promise<ApiResponse<PingResult>> {
  return request<PingResult>('POST', '/network/ping', options)
}

// postSSE 以 POST + fetch streaming 实现 SSE（EventSource 不支持 POST）
// onError：网络失败或非 2xx 响应时回调（正常流结束不触发）
function postSSE(url: string, body: unknown, onData: (data: any) => void, onError?: (err: Error) => void): EventSource {
  const session = getSessionToken()
  const controller = new AbortController()

  const reportError = (err: unknown) => {
    if (controller.signal.aborted) return
    onError?.(err instanceof Error ? err : new Error(String(err)))
  }

  fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...(session ? { 'Authorization': 'Bearer ' + session } : {}),
    },
    body: JSON.stringify(body),
    signal: controller.signal,
  }).then(async (res) => {
    if (!res.ok) {
      let msg = `HTTP ${res.status}`
      try {
        const j = await res.json()
        if (j?.message) msg = j.message
      } catch {}
      reportError(new Error(msg))
      return
    }
    const reader = res.body?.getReader()
    const decoder = new TextDecoder()
    if (!reader) {
      reportError(new Error('empty response body'))
      return
    }

    let buffer = ''
    while (true) {
      const { done, value } = await reader.read()
      if (done) break
      buffer += decoder.decode(value, { stream: true })

      const lines = buffer.split('\n')
      buffer = lines.pop() || ''

      for (const line of lines) {
        if (line.startsWith('data: ')) {
          try {
            onData(JSON.parse(line.slice(6)))
          } catch {}
        }
      }
    }
  }).catch(reportError)

  // Return a fake EventSource-like object with abort capability
  return { close: () => controller.abort() } as unknown as EventSource
}

export interface StreamError {
  type: 'error'
  message: string
}

export function networkPingStream(
  options: PingOptions,
  onPacket: (pkt: PingPacket | { type: string; sent: number; received: number; lost: number; min_rtt: number; max_rtt: number; avg_rtt: number } | StreamError) => void,
  onError?: (err: Error) => void,
): EventSource {
  return postSSE(getBaseURL() + '/api/v1/network/ping/stream', options, onPacket, onError)
}

export interface TracerouteSummary {
  type: 'summary'
  host: string
  ip: string
  hops: number
}

export function networkTracerouteStream(options: TracerouteOptions, onData: (data: TracerouteHop | TracerouteSummary | StreamError) => void, onError?: (err: Error) => void): EventSource {
  return postSSE(getBaseURL() + '/api/v1/network/traceroute/stream', options, onData, onError)
}

export function networkTraceroute(options: TracerouteOptions): Promise<ApiResponse<TracerouteResult>> {
  return request<TracerouteResult>('POST', '/network/traceroute', options)
}

export function getNetworkInterfaces(): Promise<ApiResponse<NetworkInterface[]>> {
  return request<NetworkInterface[]>('GET', '/network/interfaces')
}
